### DISCUSSION TITLE
new version 2.0 changes

### DISCUSSION THREAD

**Discussion by GitMax42** [biodiv/contactjs#23] [Q&A] [answered]
Great that contactjs is well maintained!
But as the examples on https://biodiv.github.io/contactjs/documentation/examples/ are not updated I have some difficulties upgrading to the new version. It is sad, that the non-module version is removed, but ok, as most browser support it now.
But how to migrate the swiperight / swipeleft EventListener, they are not available any more?

**Comment by biodiv** [comment]
Thanks for your feedback.
I tried to offer a non-module version. However, this turned out to be difficult with the TypeScript codebase. I think a non-module version would be nice though - maybe there is a way to produce it.

I looked into the code and it seems that the direction specific events are simply not fired for `Swipe`. A fix will be provided shortly.

**Comment by biodiv** [comment]
@ManfredGruber please try v2.0.1

**Comment by biodiv** [comment]
@ManfredGruber please check the updated codepen example:

[https://codepen.io/biodiv/pen/jOLQNQQ](https://codepen.io/biodiv/pen/jOLQNQQ)

It _should_ work now. I did not change any logic in the codepen, except the first line in javascript using the `import` statement, because the pen had used the non-module version. As the changes between 2.0.0 and 2.0.1 are extremely small: have you tried to clear your browser cache, to make sure the new file is being used?

**Reply by GitMax42** [reply]
If you open DevTools in Chrome you see an error message in the Console tab:

**contact.js:1567 Uncaught SyntaxError: Unexpected token 'export'**

**Comment by biodiv** [comment]
@ManfredGruber I managed to create a non-module version manually from the module version, which actually has been very easy.
You can download it here:  [https://github.com/biodiv/contactjs/releases/download/v2.0.1/contact.nomodule.js](https://github.com/biodiv/contactjs/releases/download/v2.0.1/contact.nomodule.js)

**Comment by biodiv** [comment]
This is weird. I tested the exact code you provided and it worked in Chromium.

If you are using a mouse for testing, please make sure that
- you move the mouse fast enough
- keep moving the mouse after you have released the mouse button

Maybe the recognition parameters need to be adjusted?

**HTML:**

```
<html>
    <head>
        <style>
            #contact {
                width: 10rem;
                height: 10rem;
                background: red;
                margin:0 auto;
                touch-action:none;
            }

            body {
                padding-top: 10rem;
                width:100%;
            }
        </style>
    </head>
    <body>
        <div id="contact">

        </div>
        <div id="log" style="text-align:center;">

        </div>
        <script type="module" src="test.js"></script>
    </body>
</html>
```

**test.js:**

```
import { Pan, PointerListener } from "https://biodiv.github.io/contactjs/assets/js/contact.js";

var domElement = document.getElementById("contact");
var log = document.getElementById("log");

var options = {
  "supportedGestures" : [Pan]
};
var pointerListener = new PointerListener(domElement, options);

function onSwipeLeft(event){
    log.textContent = log.textContent + ' swipeleft';
}
function onSwipeRight(event){
    log.textContent = log.textContent + ' swiperight';
}

function onPan(event){
    log.textContent = 'pan';
}

domElement.addEventListener('swiperight', onSwipeRight);
domElement.addEventListener('swipeleft', onSwipeLeft);
domElement.addEventListener('pan', onPan);
```

**Comment by biodiv** [comment]
The code you provided does **not** contain any errors. We have to debug this issue. Therefore, I will provide a debug version for you shortly as I cannot reproduce the error.

Below is a video of the current CodePen showing how I perform a swipe with the mouse triggering `swipeleft` and `swiperight`. 

https://user-images.githubusercontent.com/3106208/188806252-71f54cf5-b0a8-49e2-8384-6e9e4f1da944.mov

Recorded on a Mac using Safari. It also works for me in Firefox/Chromium on Linux. Nevertheless, v2.0.x should behave the same as v1.4.1. I am currently _guessing_ that you perform the motion differently than I am and somehow v2.0.1 misinterprets your input.

**Comment by GitMax42** [comment]


**Reply by biodiv** [reply]
Thanks for the pen. Yes, I can see swipe events in the console for both versions. In my case, there is no difference between the versions regarding the recognition. Tested in Safari 15.6.1 and Firefox 103.0.2. Also, i do not get erroneous pan events.

**Comment by biodiv** [comment]
@ManfredGruber Please use the following for debugging:

[contact-debug.zip](https://github.com/biodiv/contactjs/files/9504057/contact-debug.zip)

and filter the console output for `dismiss`.

You should see something like

```
[Pan] dismissing swipe. Direction: 0
[Pan] dismissing swipe. Final speed: 233.0097087378641 < 600
```
Posting your log output here would be highly appreciated. Thanks!

**Reply by SE2Dev** [reply]
> Can you provide this a no-module version (maybe already compressed)?

I was unaware that there were still a non-trivial amount of people not using bundlers or native modules in their development pipeline; I can see if there's a better way for us to produce an old-school version of the library that exposes globals (or something similar) when I have the chance.

**Reply by biodiv** [reply]
I implemented minor changes to the recognition, and activated the debug mode. However, I am a bit surprised that this fixed it. Thanks for the log files! I will release an official update soon, but first I will try to create a non-module version - by hand as I have not found an automated way yet.

**Reply by GitMax42** [reply]
@biodiv : Thank you so much, I'll test every version you provide and give you feedback!

@SE2Dev : I do my bundling of CSS and JavaScript with plain bash scripts - yes, that might be old-school but it works for over 20 years and I do not have the topic, that bundlers are showing up, disappearing or breaking the deployment because of changes / dependencies. 
The support for modules is about 95.3% which would be ok for me, but the problem in my opinion is still the way how Content-Security-Policy is checked in the browser for it (if you are not doing the extra bundling step).
Summary: I would love to see a non-module version, if this is not possible, I'll find my way around it...

**Reply by SE2Dev** [reply]
@ManfredGruber No worries! I'll see what I can do!

**Comment by biodiv** [comment]
@ManfredGruber please try the following non-module attempt. Ideally, you would test it in multiple browsers.

[contact-test-nomodule.zip](https://github.com/biodiv/contactjs/files/9507578/contact-test-nomodule.zip)

I manually edited the module version `contact.js` by replacing exports with variables. I hope @SE2Dev comes up with a better approach.

**Reply by GitMax42** [reply]
Your "gestures-nomodule" works, I see swipe events in the top right message box.

I added that no-modules version to my website and it does **not work** :-(
=> so it seems to be a problem on my side (still it works with old version v1.4.1)

By adding     "DEBUG_GESTURES" : true    to my options (I saw that in your code), I now get the following errors in my Chrome console:

______________
[pan] validating using initialParameters
index-a2.js:453 [pan] validating global min: required: 0, pointer: 0.20000004768371582
index-a2.js:453 [pan] validating global boolean: required: true, pointer: true
index-a2.js:468 validated: required value: true, current value: true
index-a2.js:453 [pan] validating live min: required: 10, pointer: 0.14453125
index-a2.js:457 [pan] invalidated 
index-a2.js:523 not firing event pan. No SinglePointerInput found
index-a2.js:433 [pan] validating using initialParameters
index-a2.js:453 [pan] validating global min: required: 0, pointer: 66.40000009536743
index-a2.js:453 [pan] validating global boolean: required: true, pointer: true
index-a2.js:468 validated: required value: true, current value: true
index-a2.js:453 [pan] validating live min: required: 10, pointer: 0.3149319432929121
index-a2.js:457 [pan] invalidated 
index-a2.js:523 not firing event pan. No SinglePointerInput found
index-a2.js:433 [pan] validating using initialParameters
index-a2.js:453 [pan] validating global min: required: 0, pointer: 83
index-a2.js:453 [pan] validating global boolean: required: true, pointer: true
index-a2.js:468 validated: required value: true, current value: true
index-a2.js:453 [pan] validating live min: required: 10, pointer: 2.9750139508601467
index-a2.js:457 [pan] invalidated 
index-a2.js:523 not firing event pan. No SinglePointerInput found
index-a2.js:433 [pan] validating using initialParameters
index-a2.js:453 [pan] validating global min: required: 0, pointer: 104.60000014305115
index-a2.js:453 [pan] validating global boolean: required: true, pointer: true
index-a2.js:468 validated: required value: true, current value: true
index-a2.js:453 [pan] validating live min: required: 10, pointer: 16.73109302696489
index-a2.js:545 [Gestures] detected and firing event panstart
index-a2.js:545 [Gestures] detected and firing event pan
index-a2.js:1599 ::: pan
index-a2.js:567 [Gestures] detected and firing event panleft
index-a2.js:835 [Pan] preventing touchmove default
index-a2.js:496 [Gesture] PointerManagerState invalidated pan: nopointer
index-a2.js:826 [Pan] dismissing swipe. Final speed: 0 < 600
index-a2.js:585 [pan] ended. Setting pan.state = inactive
index-a2.js:545 [Gestures] detected and firing event panend
______________

Can you read some information out of it? 

The only code left, where I produced the log above is:
(function () {
	'use strict';
	var options = {
		"supportedGestures" : [Pan],
		"DEBUG_GESTURES" : true
	};
	var photoSwipe = document.querySelector("#s");
	var photoPointerListener = new PointerListener(photoSwipe, options);
	photoPointerListener.a=1;
	photoSwipe.addEventListener('swiperight', function() { console.log("::: swipe right");  });
	photoSwipe.addEventListener('swipeleft',   function() { console.log("::: swipe left"); });
	photoSwipe.addEventListener('pan',            function() { console.log("::: pan"); });
}());


Same behavior on my Android chrome browser and to see the console I use the  Webview-Test  app, which shows same console output (but I can't copy it easily)
 
BTW, the working example with old version v1.4.1 is:     https://manfredgruber.at/suedbahnhotel/#p6

**Reply by GitMax42** [reply]
Hi, I updated https://manfredgruber.at/suedbahnhotel/ for version 2.0.2, maybe that helps to find the problem.
pan events are fired, swipe events not.

**Reply by biodiv** [reply]
Very nice photos :) - maybe you can add `DEBUG_GESTURES` to your live real world example at https://manfredgruber.at/suedbahnhotel/ ?

**Reply by GitMax42** [reply]
THX :-)
my new version (with contact v2.0.2) without uncompressed output and DEBUG_GESTURES is online, please notify me, when you are done, so I can compress it again.

**Reply by biodiv** [reply]
Please try adding `touch-action:none` to the class `.lb`:

```
.lb {
    display: none;
    user-select: text;
    touch-action: none;
}
```

It _seems_ like you are listening to some `touchEvent` which prevents the `pointerEvent` from being fired. I do not know if the proposed change has unwanted side-effects.

**Reply by GitMax42** [reply]
THX THAT **solves** the problem! I'll check side effects tonight, but it looks good!
Why was that not a problem with v1.4.1?

**Reply by biodiv** [reply]
v1.4.1 blocked the touch events by itself to prevent these interferences when panning/swiping. However, Chrome or something gave me a warning, that a `passive` listener should not do this. Because of #12 I made those listeners `passive`. We should re-think that strategy and maybe make the touch listeners not passive. There still is the option to prevent contact.js from handling touch events using the option `handleTouchEvents`.

**Reply by biodiv** [reply]
@ManfredGruber if you find the time, please test the following version, which contains the old handling of touch events:

[contact.js-v2.0.3-rc.zip](https://github.com/biodiv/contactjs/files/9525354/contact.js-v2.0.3-rc.zip)

**Reply by GitMax42** [reply]
@biodiv : sure, with or, without   ".lb  touch-action: none;"  ?

**Reply by biodiv** [reply]
without `touch-action: none` would be great!

**Reply by GitMax42** [reply]
it is online and works!
I know #12, because I reported it and I was because Google Page Speed reports that. But you can also see the warning in the DevTools in Chrome:
index-a9.js:1 [Violation] Added non-passive event listener to a scroll-blocking 'touchmove' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
addTouchEventListeners

**Comment by biodiv** [comment]
Indeed both warnings are valid. 
Regarding implementation of the `touch` event management I currently have the following thoughts:

- The `touchEvent` handling during pan _intends_ to block scrolling and therefore, it would make no sense to make it passive
- the cleanest implementation is using css like `touch-action:none` or `touch-action:pan-y` on specific elements that implement `swipe`or `pan`. Depending of the direction, `horizonal` or `vertical`, different css makes the most sense
- contact.js should keep the possibility to enabled and disable the touchEvent handling with the option `handleTouchEvents`

The main question is:

**Should touch handling be enabled by default?**

Enabling the touch listeners by default will lead to warnings for devs in console or analytic tools. Disabling it by default might lead to the impression that contact.js does not work - if the dev does not take care of handling touch events using css.

@ManfredGruber @SE2Dev any thoughts on if `handleTouchEvents` should be enabled or disabled by default?

**Reply by GitMax42** [reply]
good points, I would design it the safest way: enable the touch listeners by default and document clearly, that it can be disabled with the handleTouchEvents option but then needs additional CSS.

**Comment by GitMax42** [comment, marked as answer]
The solution in [v2.0.3](https://github.com/biodiv/contactjs/releases/tag/v2.0.3) works for me!  **THANK YOU @biodiv** 

Summary of my v2 upgrade:
1. the [options object](https://biodiv.github.io/contactjs/documentation/contact-js/#Options) for PointerListener needs "handleTouchEvents: false". 
2. the HTML element, which is the target for swipe events, needs a [touch-action CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action). In my case 'none' as swipe left/right/up/down is needed.

=> result: no messages ala _"Added non-passive event listener to a scroll-blocking 'touchmove' event."_ in the Chrome console and clean check result in [https://pagespeed.web.dev/](https://pagespeed.web.dev/)
